File: manifold_sampling/m/update_models.m
function [valid, Gres, Hres, X, F, h, nf, Hash] = update_models(hfun, Ffun, n, p, nf, nf_max, xkin, delta, F, X, h, Hres, fq_pars, geometry_pt_flag, Hash, tol, L, U)
    Cres = F(xkin, :);
    Res = zeros(size(F)); % Stores the residuals for model updates

    assert(ndims(Hres) == 3, 'update_models: Hres must be a 3-D array, but has ndims(Hres) = %d.', ndims(Hres)); % GH Actions debug prints
    assert(size(Hres, 3) == length(Cres), 'update_models: size(Hres,3) = %d does not match size(F,2) = %d.', size(Hres, 3), length(Cres)); % GH Actions debug prints

    for i = 1:nf
        D = X(i, :) - X(xkin, :);
        for j = 1:length(Cres)
            Res(i, j) = (F(i, j) - Cres(j)) - .5 * D * Hres(:, :, j) * D';
        end
    end
    [Mdir, np, valid, Gres, Hresdel, ~] = formquad(X(1:nf, :), Res(1:nf, :), delta, xkin, fq_pars.npmax, fq_pars.Par, 0);
    % Evaluate geometry points
    if np < n && geometry_pt_flag % Must obtain and evaluate bounded geometry points
        [Mdir, np] = bmpts(X(xkin, :), Mdir(1:n - np, :), L, U, delta, fq_pars.Par(3));
        for i = 1:min(n - np, nf_max - nf)
            [nf, X, F, h, Hash] = call_user_scripts(nf, X, F, h, Hash, Ffun, hfun, X(xkin, :) + Mdir(i, :), tol, L, U, 1);
            D = Mdir(i, :);
            for j = 1:p
                Res(nf, j) = (F(nf, j) - Cres(j)) - .5 * D * Hres(:, :, j) * D';
            end
        end
        [~, ~, valid, Gres, Hresdel, ~] = formquad(X(1:nf, :), Res(1:nf, :), delta, xkin, fq_pars.npmax, fq_pars.Par, 0);
        if isempty(Gres)
            return
        end
    end

    if ~isempty(Gres) % We'll be doing evaluations; Hres will be updated after that
        Hres = Hres + Hresdel;
    end
end

File: manifold_sampling/py/update_models.py
import numpy as np
from ibcdfo.pounders import bmpts, formquad

from .call_user_scripts import call_user_scripts


def update_models(hfun, Ffun, n, p, nf, nfmax, xkin, delta, F, X, h, Hres, fq_pars, geometry_pt_flag, Hash, tol, L, U):
    Cres = F[xkin, :]
    Res = np.zeros(F.shape)  # Stores the residuals for model updates

    D = np.zeros((nf + 1, X.shape[1]))
    D[:, :] = X[:nf + 1, :] - X[xkin]
    for i, Di in enumerate(D):
        Res[i, :] = (F[i, :] - Cres[:]) - 0.5 * np.einsum("i,ijk,j->k", Di, Hres, Di)

    Mdir, mp, valid, Gres, Hresdel, _ = formquad(X[: nf + 1, :], Res[: nf + 1, :], delta, xkin, fq_pars["npmax"], fq_pars["Par"], 0)

    # Evaluate geometry points
    if mp < n and geometry_pt_flag:  # Must obtain and evaluate bounded geometry points
        Mdir, mp = bmpts(X[xkin], Mdir[: n - mp], L, U, delta, fq_pars["Par"][2])

        for i in range(min(n - mp, nfmax - nf - 1)):
            nf, X, F, h, Hash, _ = call_user_scripts(nf, X, F, h, Hash, Ffun, hfun, X[xkin, :] + Mdir[i, :], tol, L, U, 1)
            Di = Mdir[i, :]
            Res[nf, :] = (F[nf, :] - Cres[:]) - 0.5 * np.einsum("i,ijk,j->k", Di, Hres, Di)

        _, _, valid, Gres, Hresdel, _ = formquad(X[: nf + 1], Res[: nf + 1], delta, xkin, fq_pars["npmax"], fq_pars["Par"], 0)

        if len(Gres) == 0:
            return valid, Gres, Hres, X, F, h, nf, Hash

    if len(Gres):  # We'll be doing evaluations; Hres will be updated after that
        Hres = Hres + Hresdel

    return valid, Gres, Hres, X, F, h, nf, Hash
